>From 2dd10ef70c1cb36748b04c5d9425e4b511ece969 Mon Sep 17 00:00:00 2001
From: Fabrice Fontaine <fabrice.fontaine@orange.com>
Date: Wed, 14 Mar 2012 22:37:10 +0100
Subject: [PATCH 1/6] Fix getaddrinfo() loop

Commit b116d10f did the following change:
    Use switch, int and sa_family_t with AF_INET in uri.c.

This breaks when getaddrinfo() only returns a single record, as in that
case the "break" only exits the switch statement and the loop-step
"res=res->ai_next" is still executed. After that "res == NULL" is
wrongly interpreted as not having found an AF_INET or AF_INET6 address.
---
 upnp/src/genlib/net/uri/uri.c | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/upnp/src/genlib/net/uri/uri.c b/upnp/src/genlib/net/uri/uri.c
index dff0c96..96b2a32 100644
--- upnp/src/genlib/net/uri/uri.c
+++ upnp/src/genlib/net/uri/uri.c
@@ -387,7 +387,7 @@ static int parse_hostport(
 
 			ret = getaddrinfo(srvname, NULL, &hints, &res0);
 			if (ret == 0) {
-				for (res = res0; res && !ret; res = res->ai_next) {
+				for (res = res0; res; res = res->ai_next) {
 					switch (res->ai_family) {
 					case AF_INET:
 					case AF_INET6:
@@ -395,12 +395,10 @@ static int parse_hostport(
 						memcpy(&out->IPaddress,
 						       res->ai_addr,
 						       res->ai_addrlen);
-						ret=1;
-						break;
-					default:
-						break;
+						goto found;
 					}
 				}
+found:
 				freeaddrinfo(res0);
 				if (res == NULL)
 					/* Didn't find an AF_INET or AF_INET6 address. */
-- 
2.0.0.rc0

